File: src/novel_generator/schema/chapter.py
"""Chapter schema definitions."""

from enum import Enum

from pydantic import BaseModel, Field


class EventType(str, Enum):
    """Types of events that can occur in a chapter."""

    BATTLE = "battle"
    DIALOGUE = "dialogue"
    DISCOVERY = "discovery"
    TRAINING = "training"
    ROMANCE = "romance"
    BETRAYAL = "betrayal"
    DEATH = "death"
    POWER_UP = "power_up"
    FLASHBACK = "flashback"
    CLIFFHANGER = "cliffhanger"


class ChapterEvent(BaseModel):
    """Structured event that occurred in a chapter."""

    type: EventType = Field(..., description="Type of event")
    participants: list[str] = Field(..., description="Character IDs involved")
    description: str = Field(..., description="Brief description of what happened")
    outcome: str | None = Field(default=None, description="Result of the event")
    consequences: dict[str, str] = Field(
        default_factory=dict,
        description="Consequences (e.g., {'주인공_부상': '왼팔'})",
    )


class CharacterChange(BaseModel):
    """Character state change in a chapter."""

    character_id: str = Field(..., description="Character ID")
    changes: dict[str, str] = Field(
        ..., description="What changed (e.g., {'level': '5 → 6'})"
    )


class ForeshadowingTouch(BaseModel):
    """Foreshadowing interaction in a chapter."""

    foreshadowing_id: str = Field(..., description="Foreshadowing ID")
    action: str = Field(..., description="plant, hint, or reveal")
    context: str = Field(..., description="How it was presented in the chapter")


class ChapterSummary(BaseModel):
    """Structured summary of a chapter - used for context management."""

    chapter_number: int = Field(..., description="Chapter number")
    title: str = Field(..., description="Chapter title")

    # Structured data - for retrieval
    events: list[ChapterEvent] = Field(default_factory=list)
    character_changes: list[CharacterChange] = Field(default_factory=list)
    foreshadowing_touched: list[ForeshadowingTouch] = Field(default_factory=list)

    # Text summaries - for context injection
    plot_summary: str = Field(..., description="1-2 sentence plot summary")
    emotional_beat: str = Field(..., description="Emotional tone/beat of the chapter")
    cliffhanger: str | None = Field(default=None, description="Cliffhanger if any")

    # Validation
    word_count: int = Field(default=0, description="Actual word count of chapter")
    style_score: float | None = Field(
        default=None, description="Kakao style compliance score 0-1"
    )

File: web/src/lib/schema/chapter.ts
import { z } from "zod";

// --- Enums ---

export const EventType = z.enum([
  "battle",
  "dialogue",
  "discovery",
  "training",
  "romance",
  "betrayal",
  "death",
  "power_up",
  "flashback",
  "cliffhanger",
]);
export type EventType = z.infer<typeof EventType>;

// --- Schemas ---

export const ChapterEventSchema = z.object({
  type: EventType.describe("Type of event"),
  participants: z.array(z.string()).describe("Character IDs involved"),
  description: z.string().describe("Brief description of what happened"),
  outcome: z.string().nullable().default(null).describe("Result of the event"),
  consequences: z
    .record(z.string(), z.string())
    .default({})
    .describe("Consequences (e.g., {'주인공_부상': '왼팔'})"),
});

export type ChapterEvent = z.infer<typeof ChapterEventSchema>;

export const CharacterChangeSchema = z.object({
  character_id: z.string().describe("Character ID"),
  changes: z
    .record(z.string(), z.string())
    .describe("What changed (e.g., {'level': '5 → 6'})"),
});

export type CharacterChange = z.infer<typeof CharacterChangeSchema>;

export const ForeshadowingTouchSchema = z.object({
  foreshadowing_id: z.string().describe("Foreshadowing ID"),
  action: z.string().describe("plant, hint, or reveal"),
  context: z.string().describe("How it was presented in the chapter"),
});

export type ForeshadowingTouch = z.infer<typeof ForeshadowingTouchSchema>;

export const ChapterSummarySchema = z.object({
  chapter_number: z.number().int().describe("Chapter number"),
  title: z.string().describe("Chapter title"),

  // Structured data - for retrieval
  events: z.array(ChapterEventSchema).default([]),
  character_changes: z.array(CharacterChangeSchema).default([]),
  foreshadowing_touched: z.array(ForeshadowingTouchSchema).default([]),

  // Text summaries - for context injection
  plot_summary: z.string().describe("1-2 sentence plot summary"),
  emotional_beat: z.string().describe("Emotional tone/beat of the chapter"),
  cliffhanger: z
    .string()
    .nullable()
    .default(null)
    .describe("Cliffhanger if any"),

  // Scene state at chapter end — used to enforce continuity in the next chapter
  ending_scene_state: z.object({
    time_of_day: z.string().describe("시간대: 아침/낮/저녁/밤"),
    location: z.string().describe("마지막 장면의 장소"),
    characters_present: z.array(z.string()).describe("마지막 장면에 있던 캐릭터 이름"),
    ongoing_action: z.string().describe("진행 중이던 행동/상황 (예: 쪽지를 꺼내려는 순간)"),
    unresolved_tension: z.string().describe("아직 해소되지 않은 긴장/질문"),
  }).nullable().default(null).describe("화 마지막 시점의 장면 상태 — 다음 화 연속성 보장용"),

  // Validation
  word_count: z
    .number()
    .int()
    .default(0)
    .describe("Actual word count of chapter"),
  style_score: z
    .number()
    .nullable()
    .default(null)
    .describe("Kakao style compliance score 0-1"),
});

export type ChapterSummary = z.infer<typeof ChapterSummarySchema>;
